File: sdk/go/pkg/a2areg/client_test.go
package a2areg

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestNewA2ARegClient(t *testing.T) {
	tests := []struct {
		name string
		opts A2ARegClientOptions
		want string
	}{
		{
			name: "default options",
			opts: DefaultOptions(),
			want: "http://localhost:8000",
		},
		{
			name: "custom options",
			opts: A2ARegClientOptions{
				RegistryURL:  "https://registry.example.com",
				ClientID:     "test-client",
				ClientSecret: "test-secret",
				Timeout:      60 * time.Second,
				APIKey:       "test-key",
				Scope:        "read write admin",
			},
			want: "https://registry.example.com",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			client := NewA2ARegClient(tt.opts)
			assert.Equal(t, tt.want, client.registryURL)
			assert.NotNil(t, client.httpClient)
		})
	}
}

func TestA2ARegClient_GetHealth(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/health", r.URL.Path)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"status":  "healthy",
			"version": "1.0.0",
		})
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	health, err := client.GetHealth()
	require.NoError(t, err)
	assert.Equal(t, "healthy", health["status"])
	assert.Equal(t, "1.0.0", health["version"])
}

func TestA2ARegClient_Authenticate(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/auth/oauth/token", r.URL.Path)
		assert.Equal(t, "POST", r.Method)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"access_token": "test-token",
			"expires_in":   3600,
		})
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL:  server.URL,
		ClientID:     "test-client",
		ClientSecret: "test-secret",
	})

	err := client.Authenticate()
	require.NoError(t, err)
	assert.Equal(t, "test-token", client.accessToken)
	assert.NotNil(t, client.tokenExpiresAt)
}

func TestA2ARegClient_Authenticate_WithAPIKey(t *testing.T) {
	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: "http://localhost:8000",
		APIKey:      "test-key",
	})

	err := client.Authenticate()
	require.NoError(t, err)
	// Should not make any requests
	assert.Equal(t, "test-key", client.apiKey)
}

func TestA2ARegClient_Authenticate_MissingCredentials(t *testing.T) {
	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: "http://localhost:8000",
	})

	err := client.Authenticate()
	assert.Error(t, err)
	assert.IsType(t, &AuthenticationError{}, err)
}

func TestA2ARegClient_ListAgents(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/agents/public", r.URL.Path)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"agents": []map[string]interface{}{
				{"id": "agent-1", "name": "Test Agent"},
			},
			"total": 1,
			"page":  1,
			"limit": 20,
		})
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	result, err := client.ListAgents(1, 20, true)
	require.NoError(t, err)
	agents, _ := result["agents"].([]interface{})
	assert.Len(t, agents, 1)
}

func TestA2ARegClient_GetAgent(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/agents/agent-1", r.URL.Path)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"id":          "agent-1",
			"name":        "Test Agent",
			"description": "A test agent",
			"version":     "1.0.0",
			"provider":    "test-provider",
		})
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	agent, err := client.GetAgent("agent-1")
	require.NoError(t, err)
	assert.Equal(t, "agent-1", *agent.ID)
	assert.Equal(t, "Test Agent", agent.Name)
}

func TestA2ARegClient_GetAgent_NotFound(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	_, err := client.GetAgent("nonexistent")
	assert.Error(t, err)
	assert.IsType(t, &NotFoundError{}, err)
}

func TestA2ARegClient_PublishAgent(t *testing.T) {
	publishServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/agents/publish" {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusOK)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"agentId": "agent-123",
			})
		} else if r.URL.Path == "/agents/agent-123" {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusOK)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"id":          "agent-123",
				"name":        "New Agent",
				"description": "A new agent",
				"version":     "1.0.0",
				"provider":    "test-provider",
			})
		}
	}))
	defer publishServer.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: publishServer.URL,
		APIKey:      "test-key",
	})

	agent := &Agent{
		Name:        "New Agent",
		Description: "A new agent",
		Version:     "1.0.0",
		Provider:    "test-provider",
		IsPublic:    true,
	}

	published, err := client.PublishAgent(agent, false)
	require.NoError(t, err)
	assert.Equal(t, "agent-123", *published.ID)
	assert.Equal(t, "New Agent", published.Name)
}

func TestA2ARegClient_ValidateAgent(t *testing.T) {
	client := NewA2ARegClient(DefaultOptions())

	tests := []struct {
		name    string
		agent   *Agent
		wantErr bool
		errType string
	}{
		{
			name: "valid agent",
			agent: &Agent{
				Name:        "Test Agent",
				Description: "A test agent",
				Version:     "1.0.0",
				Provider:    "test-provider",
			},
			wantErr: false,
		},
		{
			name: "missing name",
			agent: &Agent{
				Description: "A test agent",
				Version:     "1.0.0",
				Provider:    "test-provider",
			},
			wantErr: true,
			errType: "ValidationError",
		},
		{
			name: "missing description",
			agent: &Agent{
				Name:     "Test Agent",
				Version:  "1.0.0",
				Provider: "test-provider",
			},
			wantErr: true,
			errType: "ValidationError",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := client.ValidateAgent(tt.agent)
			if tt.wantErr {
				assert.Error(t, err)
				assert.IsType(t, &ValidationError{}, err)
			} else {
				assert.NoError(t, err)
			}
		})
	}
}

func TestA2ARegClient_ErrorHandling(t *testing.T) {
	tests := []struct {
		name       string
		statusCode int
		errType    interface{}
	}{
		{
			name:       "401 Unauthorized",
			statusCode: http.StatusUnauthorized,
			errType:    &AuthenticationError{},
		},
		{
			name:       "403 Forbidden",
			statusCode: http.StatusForbidden,
			errType:    &AuthenticationError{},
		},
		{
			name:       "404 Not Found",
			statusCode: http.StatusNotFound,
			errType:    &NotFoundError{},
		},
		{
			name:       "422 Unprocessable Entity",
			statusCode: http.StatusUnprocessableEntity,
			errType:    &ValidationError{},
		},
		{
			name:       "500 Internal Server Error",
			statusCode: http.StatusInternalServerError,
			errType:    &A2AError{},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				w.WriteHeader(tt.statusCode)
			}))
			defer server.Close()

			client := NewA2ARegClient(A2ARegClientOptions{
				RegistryURL: server.URL,
				APIKey:      "test-key",
			})

			_, err := client.GetAgent("agent-1")
			assert.Error(t, err)
			assert.IsType(t, tt.errType, err)
		})
	}
}

func TestA2ARegClient_SetAPIKey(t *testing.T) {
	client := NewA2ARegClient(DefaultOptions())
	client.SetAPIKey("new-key")
	assert.Equal(t, "new-key", client.apiKey)
}

func TestA2ARegClient_SearchAgents(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/agents/search", r.URL.Path)
		assert.Equal(t, "POST", r.Method)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"agents": []map[string]interface{}{
				{"id": "agent-1", "name": "Recipe Agent"},
			},
			"total": 1,
		})
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	result, err := client.SearchAgents("recipe", map[string]interface{}{"tags": []string{"cooking"}}, false, 1, 20)
	require.NoError(t, err)
	agents, _ := result["agents"].([]interface{})
	assert.Len(t, agents, 1)
}

func TestA2ARegClient_GenerateAPIKey(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/security/api-keys", r.URL.Path)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"api_key":    "generated-key",
			"key_id":     "key-123",
			"scopes":     []string{"read", "write"},
			"created_at": "2024-01-01T00:00:00Z",
		})
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	apiKey, keyInfo, err := client.GenerateAPIKey([]string{"read", "write"}, nil)
	require.NoError(t, err)
	assert.Equal(t, "generated-key", apiKey)
	assert.Equal(t, "key-123", keyInfo["key_id"])
}

func TestA2ARegClient_ValidateAPIKey(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"key_id": "key-123",
			"scopes": []string{"read", "write"},
			"active": true,
		})
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	result, err := client.ValidateAPIKey("test-key", []string{"read"})
	require.NoError(t, err)
	assert.NotNil(t, result)
	assert.Equal(t, "key-123", result["key_id"])
}

func TestA2ARegClient_RevokeAPIKey(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	revoked, err := client.RevokeAPIKey("key-123")
	require.NoError(t, err)
	assert.True(t, revoked)
}

func TestA2ARegClient_ListAPIKeys(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode([]map[string]interface{}{
			{
				"key_id":    "key-1",
				"scopes":    []string{"read"},
				"created_at": "2024-01-01T00:00:00Z",
			},
		})
	}))
	defer server.Close()

	client := NewA2ARegClient(A2ARegClientOptions{
		RegistryURL: server.URL,
		APIKey:      "test-key",
	})

	keys, err := client.ListAPIKeys(true)
	require.NoError(t, err)
	assert.Len(t, keys, 1)
}

File: sdk/typescript/tests/client.test.ts
/**
 * Tests for A2ARegClient
 */

import { A2ARegClient } from '../src/client';
import { agentFromDict } from '../src/models';
import { A2AError, AuthenticationError, ValidationError, NotFoundError } from '../src/exceptions';

// Mock fetch globally
(global as any).fetch = jest.fn();

describe('A2ARegClient', () => {
  let client: A2ARegClient;
  const mockFetch = (global as any).fetch as jest.Mock;

  beforeEach(() => {
    client = new A2ARegClient({
      registryUrl: 'https://registry.example.com',
      apiKey: 'test-api-key',
    });
    mockFetch.mockClear();
    // Reset fetch mock
    (global as any).fetch = mockFetch;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Initialization', () => {
    it('should initialize with default values', () => {
      const defaultClient = new A2ARegClient();
      expect(defaultClient).toBeInstanceOf(A2ARegClient);
    });

    it('should initialize with custom options', () => {
      const customClient = new A2ARegClient({
        registryUrl: 'https://custom.example.com',
        clientId: 'client-id',
        clientSecret: 'client-secret',
        timeout: 60,
        scope: 'read write admin',
      });
      expect(customClient).toBeInstanceOf(A2ARegClient);
    });

    it('should set API key', () => {
      client.setApiKey('new-api-key');
      expect(client).toBeInstanceOf(A2ARegClient);
    });
  });

  describe('Authentication', () => {
    it('should authenticate with API key (skip OAuth)', async () => {
      const apiKeyClient = new A2ARegClient({
        registryUrl: 'https://registry.example.com',
        apiKey: 'test-key',
      });

      await apiKeyClient.authenticate();
      // Should not make any fetch calls since API key is set
      expect(mockFetch).not.toHaveBeenCalled();
    });

    it('should authenticate with OAuth', async () => {
      const oauthClient = new A2ARegClient({
        registryUrl: 'https://registry.example.com',
        clientId: 'test-client',
        clientSecret: 'test-secret',
      });

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          access_token: 'test-token',
          expires_in: 3600,
        }),
      });

      await oauthClient.authenticate();
      expect(mockFetch).toHaveBeenCalledWith(
        'https://registry.example.com/auth/oauth/token',
        expect.objectContaining({
          method: 'POST',
        })
      );
    });

    it('should throw AuthenticationError on OAuth failure', async () => {
      const oauthClient = new A2ARegClient({
        registryUrl: 'https://registry.example.com',
        clientId: 'test-client',
        clientSecret: 'test-secret',
      });

      mockFetch.mockResolvedValueOnce({
        ok: false,
        statusText: 'Unauthorized',
      });

      await expect(oauthClient.authenticate()).rejects.toThrow(AuthenticationError);
    });

    it('should throw error when client credentials missing', async () => {
      const noCredsClient = new A2ARegClient({
        registryUrl: 'https://registry.example.com',
      });

      await expect(noCredsClient.authenticate()).rejects.toThrow(AuthenticationError);
    });
  });

  describe('Health', () => {
    it('should get health status', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ status: 'healthy', version: '1.0.0' }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const health = await client.getHealth();
      expect(health).toEqual({ status: 'healthy', version: '1.0.0' });
      expect(mockFetch).toHaveBeenCalledWith(
        'https://registry.example.com/health',
        expect.any(Object)
      );
    });
  });

  describe('Agents', () => {
    it('should list public agents', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          agents: [{ id: 'agent-1', name: 'Test Agent' }],
          total: 1,
          page: 1,
          limit: 20,
        }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const result = await client.listAgents(1, 20, true);
      expect(result.agents).toHaveLength(1);
      expect(result.agents[0].id).toBe('agent-1');
    });

    it('should list entitled agents', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          agents: [{ id: 'agent-2', name: 'Private Agent' }],
          total: 1,
          page: 1,
          limit: 20,
        }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const result = await client.listAgents(1, 20, false);
      expect(result.agents).toHaveLength(1);
      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining('/agents/entitled'),
        expect.any(Object)
      );
    });

    it('should get agent by ID', async () => {
      const agentData = {
        id: 'agent-1',
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => agentData,
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const agent = await client.getAgent('agent-1');
      expect(agent.id).toBe('agent-1');
      expect(agent.name).toBe('Test Agent');
    });

    it('should throw NotFoundError when agent not found', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 404,
        statusText: 'Not Found',
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      await expect(client.getAgent('nonexistent')).rejects.toThrow(NotFoundError);
    });

    it('should get agent card', async () => {
      const cardData = {
        name: 'Test Agent',
        description: 'A test agent',
        url: 'https://test.example.com',
        version: '1.0.0',
        capabilities: { streaming: false },
        securitySchemes: [{ type: 'apiKey' }],
        skills: [],
        interface: {
          preferredTransport: 'jsonrpc',
          defaultInputModes: ['text/plain'],
          defaultOutputModes: ['text/plain'],
        },
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => cardData,
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const card = await client.getAgentCard('agent-1');
      expect(card.name).toBe('Test Agent');
      expect(card.version).toBe('1.0.0');
    });
  });

  describe('Search', () => {
    it('should search agents', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          agents: [{ id: 'agent-1', name: 'Recipe Agent' }],
          total: 1,
          page: 1,
          limit: 20,
        }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const results = await client.searchAgents('recipe', { tags: ['cooking'] });
      expect(results.agents).toHaveLength(1);
      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining('/agents/search'),
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({
            query: 'recipe',
            filters: { tags: ['cooking'] },
            semantic: false,
            page: 1,
            limit: 20,
          }),
        })
      );
    });

    it('should search with semantic search', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ agents: [], total: 0 }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      await client.searchAgents('recipe', undefined, true);
      const callArgs = (mockFetch.mock.calls[0][1] as any).body;
      const body = JSON.parse(callArgs);
      expect(body.semantic).toBe(true);
    });
  });

  describe('Publish Agent', () => {
    it('should publish agent successfully', async () => {
      const agentData = {
        name: 'New Agent',
        description: 'A new agent',
        version: '1.0.0',
        provider: 'test-provider',
        location_url: 'https://api.example.com/agent',
        is_public: true,
      };

      // Mock publish response
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ agentId: 'agent-123' }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      // Mock get agent response
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ ...agentData, id: 'agent-123' }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const published = await client.publishAgent(agentData);
      expect(published.id).toBe('agent-123');
      expect(published.name).toBe('New Agent');
    });

    it('should publish agent with validation', async () => {
      const validAgent = {
        name: 'Valid Agent',
        description: 'A valid agent',
        version: '1.0.0',
        provider: 'test-provider',
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ agentId: 'agent-123' }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ ...validAgent, id: 'agent-123' }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const published = await client.publishAgent(validAgent, true);
      expect(published).toBeDefined();
    });

    it('should throw ValidationError on invalid agent', async () => {
      const invalidAgent = {
        name: '',
        description: 'Missing name',
        version: '1.0.0',
        provider: 'test-provider',
      };

      await expect(client.publishAgent(invalidAgent, true)).rejects.toThrow(ValidationError);
    });
  });

  describe('Update Agent', () => {
    it('should update agent', async () => {
      const updatedData = {
        name: 'Updated Agent',
        description: 'Updated description',
        version: '1.1.0',
        provider: 'test-provider',
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ ...updatedData, id: 'agent-1' }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const updated = await client.updateAgent('agent-1', updatedData);
      expect(updated.name).toBe('Updated Agent');
      expect(updated.version).toBe('1.1.0');
    });
  });

  describe('Delete Agent', () => {
    it('should delete agent', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      await client.deleteAgent('agent-1');
      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining('/agents/agent-1'),
        expect.objectContaining({
          method: 'DELETE',
        })
      );
    });
  });

  describe('Validate Agent', () => {
    it('should validate valid agent', () => {
      const validAgent = agentFromDict({
        name: 'Valid Agent',
        description: 'A valid agent',
        version: '1.0.0',
        provider: 'test-provider',
      });

      const errors = client.validateAgent(validAgent);
      expect(errors).toHaveLength(0);
    });

    it('should return errors for invalid agent', () => {
      const invalidAgent = agentFromDict({
        name: '',
        description: 'Missing name',
        version: '1.0.0',
        provider: 'test-provider',
      });

      const errors = client.validateAgent(invalidAgent);
      expect(errors.length).toBeGreaterThan(0);
      expect(errors.some((e) => e.includes('name is required'))).toBe(true);
    });

    it('should validate agent card if present', () => {
      const agent = agentFromDict({
        name: 'Test Agent',
        description: 'A test agent',
        version: '1.0.0',
        provider: 'test-provider',
        agent_card: {
          name: '',
          description: 'Missing card name',
          version: '1.0.0',
          url: 'https://test.com',
          capabilities: {},
          securitySchemes: [],
          skills: [],
          interface: {
            preferredTransport: 'jsonrpc',
            defaultInputModes: ['text/plain'],
            defaultOutputModes: ['text/plain'],
          },
        },
      });

      const errors = client.validateAgent(agent);
      expect(errors.some((e) => e.includes('Agent card name is required'))).toBe(true);
    });
  });

  describe('API Key Management', () => {
    it('should generate API key', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          api_key: 'generated-key',
          key_id: 'key-123',
          scopes: ['read', 'write'],
          created_at: '2024-01-01T00:00:00Z',
        }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const result = await client.generateApiKey(['read', 'write'], 30);
      expect(result.apiKey).toBe('generated-key');
      expect(result.keyInfo.key_id).toBe('key-123');
    });

    it('should generate and authenticate with API key', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          api_key: 'new-key',
          key_id: 'key-456',
          scopes: ['read'],
          created_at: '2024-01-01T00:00:00Z',
        }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const result = await client.generateApiKeyAndAuthenticate(['read']);
      expect(result.apiKey).toBe('new-key');
    });

    it('should validate API key', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          key_id: 'key-123',
          scopes: ['read', 'write'],
          active: true,
        }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const result = await client.validateApiKey('test-key', ['read']);
      expect(result).toBeDefined();
      expect(result?.key_id).toBe('key-123');
    });

    it('should return null for invalid API key', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 401,
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const result = await client.validateApiKey('invalid-key');
      expect(result).toBeNull();
    });

    it('should revoke API key', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const result = await client.revokeApiKey('key-123');
      expect(result).toBe(true);
    });

    it('should return false when revoking nonexistent key', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 404,
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const result = await client.revokeApiKey('nonexistent-key');
      expect(result).toBe(false);
    });

    it('should list API keys', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => [
          {
            key_id: 'key-1',
            scopes: ['read'],
            created_at: '2024-01-01T00:00:00Z',
          },
        ],
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const keys = await client.listApiKeys(true);
      expect(Array.isArray(keys)).toBe(true);
      expect(keys.length).toBeGreaterThan(0);
    });
  });

  describe('Error Handling', () => {
    it('should throw AuthenticationError on 401', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 401,
        statusText: 'Unauthorized',
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      await expect(client.getAgent('agent-1')).rejects.toThrow(AuthenticationError);
    });

    it('should throw AuthenticationError on 403', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 403,
        statusText: 'Forbidden',
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      await expect(client.getAgent('agent-1')).rejects.toThrow(AuthenticationError);
    });

    it('should throw ValidationError on 422', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 422,
        statusText: 'Unprocessable Entity',
        json: async () => ({ detail: 'Validation failed' }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      await expect(client.publishAgent({} as any)).rejects.toThrow(ValidationError);
    });

    it('should throw A2AError on generic errors', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 500,
        statusText: 'Internal Server Error',
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      await expect(client.getAgent('agent-1')).rejects.toThrow(A2AError);
    });

    it('should handle request timeout', async () => {
      const timeoutClient = new A2ARegClient({
        registryUrl: 'https://registry.example.com',
        apiKey: 'test-key',
        timeout: 1,
      });

      // Mock AbortController to simulate timeout
      const originalAbortController = global.AbortController;
      (global as any).AbortController = jest.fn().mockImplementation(() => {
        const controller = new originalAbortController();
        const originalAbort = controller.abort.bind(controller);
        // Immediately abort to simulate timeout
        setTimeout(() => originalAbort(), 0);
        return controller;
      });

      mockFetch.mockImplementationOnce(() => {
        return new Promise((_resolve, reject) => {
          setTimeout(() => {
            const error = new Error('The operation was aborted');
            (error as any).name = 'AbortError';
            reject(error);
          }, 10);
        });
      });

      await expect(timeoutClient.getHealth()).rejects.toThrow(A2AError);

      // Restore original
      (global as any).AbortController = originalAbortController;
    });
  });

  describe('Registry Stats', () => {
    it('should get registry stats', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          total_agents: 100,
          public_agents: 80,
          private_agents: 20,
        }),
        headers: new Headers({ 'content-type': 'application/json' }),
      });

      const stats = await client.getRegistryStats();
      expect(stats.total_agents).toBe(100);
    });
  });
});

